pkg/binpack/efficiency.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binpack
+
+import (
+	"math"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+)
+
+// PackingEfficiency represents result packing efficiency per resource type. Computed as the total
+// resources used divided by total capacity.
+type PackingEfficiency struct {
+	NodeName string
+	CPU      float64
+	Memory   float64
+	GPU      float64
+}
+
+// LessThan compares two packing efficiencies. For a single packing we take the highest of the
+// resources' efficiency. For example, when CPU is at 0.81 and Memory is at 0.54 the avg efficiency
+// is 0.81. One packing efficiency is deemed less efficient when its avg efficiency is lower than
+// the other's packing efficiency.
+func (p *PackingEfficiency) LessThan(o PackingEfficiency) bool {
+	// TODO: GPU is explicitly excluded for now but worthwhile to reconsider in future
```

Comment:
I propose to represent a missing GPU with a different value (0 or -1), so that it does not affect this calculation and we don't ever need to come back to it.

Will send an update once I consider different options.